File: src/main/java/com/hileco/cortex/analysis/edges/EdgeFlowMapping.java
package com.hileco.cortex.analysis.edges;

import com.hileco.cortex.analysis.GraphBlock;
import com.hileco.cortex.analysis.GraphNode;
import lombok.Getter;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

@Getter
public class EdgeFlowMapping implements Edge {
    public static final EdgeUtility<EdgeFlowMapping> UTIL = new EdgeUtility<>(EdgeFlowMapping.class);

    private final Map<Integer, Set<EdgeFlow>> flowsFromSource;
    private final Map<Integer, Set<EdgeFlow>> flowsToTarget;
    private final Map<Integer, GraphBlock> blockLineMapping;
    private final Map<Integer, GraphNode> nodeLineMapping;

    public EdgeFlowMapping() {
        this.flowsFromSource = new HashMap<>();
        this.flowsToTarget = new HashMap<>();
        this.blockLineMapping = new HashMap<>();
        this.nodeLineMapping = new HashMap<>();
    }

    public void putLineMapping(Integer key, GraphBlock value) {
        this.blockLineMapping.put(key, value);
    }

    public void putLineMapping(Integer key, GraphNode value) {
        this.nodeLineMapping.put(key, value);
    }

    public void map(EdgeFlow edgeFlow) {
        this.flowsFromSource.computeIfAbsent(edgeFlow.getSource(), ignore -> new HashSet<>()).add(edgeFlow);
        this.flowsToTarget.computeIfAbsent(edgeFlow.getTarget(), ignore -> new HashSet<>()).add(edgeFlow);
    }
}

File: src/main/java/com/hileco/cortex/analysis/edges/EdgeFlowMapping.kt
package com.hileco.cortex.analysis.edges

import com.hileco.cortex.analysis.GraphBlock
import com.hileco.cortex.analysis.GraphNode
import java.util.*
import kotlin.collections.HashMap

class EdgeFlowMapping(val flowsFromSource: MutableMap<Int?, MutableSet<EdgeFlow>> = HashMap(),
                      val flowsToTarget: MutableMap<Int?, MutableSet<EdgeFlow>> = HashMap(),
                      val blockLineMapping: MutableMap<Int, GraphBlock> = HashMap(),
                      val nodeLineMapping: MutableMap<Int, GraphNode> = HashMap()) : Edge {
    fun putLineMapping(key: Int, value: GraphBlock) {
        blockLineMapping[key] = value
    }

    fun putLineMapping(key: Int, value: GraphNode) {
        nodeLineMapping[key] = value
    }

    fun map(edgeFlow: EdgeFlow) {
        flowsFromSource.computeIfAbsent(edgeFlow.source) { HashSet() }.add(edgeFlow)
        flowsToTarget.computeIfAbsent(edgeFlow.target) { HashSet() }.add(edgeFlow)
    }

    companion object {
        val UTIL = EdgeUtility(EdgeFlowMapping::class.java)
    }
}
